### PR TITLE
[Merged by Bors] - Add despawn_recursive to EntityMut

### PR DESCRIPTION
**Pull request by TheRawMeatball**

### CODE REVIEW

**Review by alice-i-cecile — APPROVED** [review]

File: crates/bevy_transform/src/hierarchy/hierarchy.rs [hunk 1]
```diff
@@ -55,6 +55,17 @@ impl<'w, 's, 'a> DespawnRecursiveExt for EntityCommands<'w, 's, 'a> {
     }
 }
 
+impl<'w> DespawnRecursiveExt for EntityMut<'w> {
+    /// Despawns the provided entity and its children.
+    fn despawn_recursive(&mut self) {
+        let entity = self.id();
+        unsafe {
+            despawn_with_children_recursive(self.world_mut(), entity);
```
Comment (payload) [inline crates/bevy_transform/src/hierarchy/hierarchy.rs:63]: Could you make the unsafe scope minimal by using instead.                `despawn_with_children_recursive(unsafe { self.world_mut() }, entity);` ?

**Review by payload — CHANGES_REQUESTED** [review]
Could you make the scope which allows unsafe operations minimal? I s

File: crates/bevy_transform/src/hierarchy/hierarchy.rs (hunk 1 above)
Comment (TheRawMeatball) [reply crates/bevy_transform/src/hierarchy/hierarchy.rs:63]: I could, but I prefer to have the unsafe block also cover the part that makes it sound.

Comment (payload) [reply crates/bevy_transform/src/hierarchy/hierarchy.rs:63]: I reread despawn_with_children_recursive again and it is calling despawning this entity itself too. Calling update_location after that would reveal the error here. There is no location of this entity anymore in this world and its .unwrap() should panic in this case.

Comment (TheRawMeatball) [reply crates/bevy_transform/src/hierarchy/hierarchy.rs:63]: Ah, of course. This also shows the semantic problem with the trait - I'll update it to take `self` instead.

**Review by inodentry — APPROVED** [review]

File: crates/bevy_transform/src/hierarchy/hierarchy.rs [hunk 2]
```diff
@@ -44,17 +44,27 @@ impl Command for DespawnRecursive {
...
         let entity = self.id();
         self.commands().add(DespawnRecursive { entity });
     }
 }
 
+impl<'w> DespawnRecursiveExt for EntityMut<'w> {
+    /// Despawns the provided entity and its children.
+    fn despawn_recursive(mut self) {
+        let entity = self.id();
+        unsafe {
```
Comment (cart) [inline crates/bevy_transform/src/hierarchy/hierarchy.rs:62]: This should have a `// SAFE: rationale` comment above it explaining why this is safe in relationship to the requirements of `self.world_mut()` (ex: EntityMut is consumed so even though the location is no longer valid, it cannot be accessed again with the invalid location).

Comment (TheRawMeatball) [reply crates/bevy_transform/src/hierarchy/hierarchy.rs:62]: I'll just copy that rationale in.

**Pull request closed without merging**